src/main/scala/io/picnicml/doddlemodel/cluster/DBSCAN.scala

```diff
@@ -0,0 +1,98 @@
+package io.picnicml.doddlemodel.cluster
+
+import breeze.linalg.functions.euclideanDistance
+import cats.syntax.option._
+import io.picnicml.doddlemodel.data.Features
+import io.picnicml.doddlemodel.typeclasses.Clusterer
+
+/** An immutable DBSCAN model.
+  *
+  * @param eps: the maximum distance between points in a group
+  * @param min_samples: the minimum number of point in a core group
+  *
+  * Examples:
+  * val model = DBSCAN()
+  * val model = DBSCAN(eps = 1.5)
+  * val model = DBSCAN(min_samples = 3)
+  * val model = DBSCAN(eps = 2.0, min_samples = 3)
+  */
+case class DBSCAN private(eps: Double, min_samples: Int,
+  private val label: Option[Array[Int]]) {}
+
+object DBSCAN {
+
+  def apply(): DBSCAN = DBSCAN(1.0, 1, none)
+
+  def apply(eps: Double): DBSCAN = {
+    require(eps > 0, "maximum distance need to be larger than 0")
+    DBSCAN(eps, 1, none)
+  }
+  def apply(min_samples: Int): DBSCAN = {
+    require(min_samples > 0,
+      "minimum number of points in a group need to be larger than 0")
+    DBSCAN(1.0, min_samples, none)
+  }
+  def apply(eps: Double, min_samples: Int): DBSCAN = {
+    require(eps > 0, "maximum distance need to be larger than 0")
+    require(min_samples > 0,
+      "minimum number of points in a group need to be larger than 0")
+    DBSCAN(eps, min_samples, none)
+  }
+
+  implicit lazy val ev: Clusterer[DBSCAN] = new Clusterer[DBSCAN] {
+
+    override def isFitted(model: DBSCAN): Boolean = model.label.isDefined
+
+    override protected def labelSafe(model: DBSCAN): Array[Int] = model.label.get
+
+    override protected def copy(model: DBSCAN): DBSCAN =
+      model.copy()
+
+    override protected def copy(model: DBSCAN, label: Array[Int]): DBSCAN =
+      model.copy(label = label.some)
+
+    override protected def fitSafe(model: DBSCAN, x: Features): DBSCAN = {
+      val xSize = x.rows
+      val label = Array.fill[Int](xSize)(Int.MaxValue)
+      var groupStartId = 0
+      var groupId = 0
+      while (groupStartId < xSize) {
+        var groupQueue = Set(groupStartId)
+        var groupCount = 1
+        label(groupStartId) = groupId
+        while (groupQueue.size > 0) {
+          val tmpGroupQueue = groupQueue
+          groupQueue = Set[Int]()
+          for (pointId <- tmpGroupQueue; i <- 0 until xSize
+            if label(i) > groupId &&
+              euclideanDistance(x(i, ::).t, x(pointId, ::).t) < model.eps) {
+            label(i) = groupId
+            groupQueue += i
+            groupCount += 1
+          }
+        }
+        if (groupCount < model.min_samples) {
```

Comment:
I think this check should also be present somewhere sooner. Consider the following points:
```scala
val x = DenseMatrix(List(0.0, 12.0), List(0.0, 9.0), List(0.0, 6.0), List(0.0, 3.0), List(0.0, 0.0))
```
i.e. 5 points spread through the `y` axis, each being 3 units apart.
Assuming the changed definition of Eps-neighborhood (`<=` instead of `<`), I think the points are not being clustered correctly for the following setting:
```scala
val model = DBSCAN(3.0, 4)
```

They all get assigned to cluster `0`, whereas they should be marked as noise (according to the scikit-learn implementation). I think the problem is that this check should be performed before doing the depth-first search, i.e. that each cluster needs some core point, a point that has atleast `minSamples` immediate neighbors.

Reply:
You won't be able to tell if all points will be assigned into one cluster before the DFS.
But I will add another test to check the `<=` instead of `<` part.

Reply:
The way I understood the algorithm from the paper is that first, you check if current point has at least `minSamples` points in its Eps-neighborhood. Then:
- if this condition is **not** fullfilled, label this point as **noise** and move on to the next point
- otherwise, this point is a core point and a new cluster will be formed around it. In this case, you do the DFS on the other points in Eps-neighborhood, propagating the ID of current cluster

Please let me know If I misunderstood something and I'll take another look

Reply:
You are absolutely correct and I was wrong. I will make another modification.